src/node_osrm.cpp

```diff
@@ -4,594 +4,507 @@
-    }
+    static void route(const Nan::FunctionCallbackInfo<v8::Value> &args);
+    static void nearest(const Nan::FunctionCallbackInfo<v8::Value> &args);
+    static void table(const Nan::FunctionCallbackInfo<v8::Value> &args);
+    static void match(const Nan::FunctionCallbackInfo<v8::Value> &args);
+    static void trip(const Nan::FunctionCallbackInfo<v8::Value> &args);
 
-    Local<Array> coordinate_pair = Local<Array>::Cast(coordinate);
-    if (coordinate_pair->Length() != 2) {
-        NanThrowError("first argument must be an array of lat, long");
-        NanReturnUndefined();
+    static void Run(const Nan::FunctionCallbackInfo<v8::Value> &args, route_parameters_ptr);
+    static void AsyncRun(uv_work_t *);
+    static void AfterRun(uv_work_t *);
+
+  private:
+    Engine(LibOSRMConfig &lib_config)
+        : Nan::ObjectWrap(), this_(make_unique<OSRM>(lib_config))
+    {
     }
 
-    route_parameters_ptr params = make_unique<RouteParameters>();
+    static Nan::Persistent<v8::Function> constructor;
+    osrm_ptr this_;
+};
 
-    params->service = "locate";
-    params->coordinates.emplace_back(static_cast<int>(coordinate_pair->Get(0)->NumberValue()*COORDINATE_PRECISION),
-                                     static_cast<int>(coordinate_pair->Get(1)->NumberValue()*COORDINATE_PRECISION));
+Nan::Persistent<v8::Function> Engine::constructor;
 
-    Run(args, std::move(params));
-    NanReturnUndefined();
+void Engine::Initialize(v8::Handle<v8::Object> target)
+{
+    v8::Local<v8::FunctionTemplate> function_template = Nan::New<v8::FunctionTemplate>(Engine::New);
+
+    function_template->InstanceTemplate()->SetInternalFieldCount(1);
+    function_template->SetClassName(Nan::New("OSRM").ToLocalChecked());
+
+    SetPrototypeMethod(function_template, "route", route);
+    SetPrototypeMethod(function_template, "nearest", nearest);
+    SetPrototypeMethod(function_template, "table", table);
+    SetPrototypeMethod(function_template, "match", match);
+    SetPrototypeMethod(function_template, "trip", trip);
+
+    constructor.Reset(function_template->GetFunction());
+    Nan::Set(target, Nan::New("OSRM").ToLocalChecked(), function_template->GetFunction());
 }
 
-NAN_METHOD(Engine::match)
+void Engine::New(const Nan::FunctionCallbackInfo<v8::Value> &args)
 {
-    NanScope();
-    if (args.Length() < 2) {
-        NanThrowTypeError("two arguments required");
-        NanReturnUndefined();
+    if (!args.IsConstructCall())
+    {
+        Nan::ThrowTypeError("Cannot call constructor as function, you need to use 'new' "
+                            "keyword");
+        return;
     }
 
-    if (args[0]->IsNull() || args[0]->IsUndefined()) {
-        NanThrowError("first arg must be an object");
-        NanReturnUndefined();
+    try
+    {
+        auto lib_config_ptr = argumentsToLibOSRMConfig(args);
+        if (!lib_config_ptr)
+        {
+            return;
+        }
+        auto engine_ptr = new Engine(*lib_config_ptr);
```

Comment:
`Engine` would definitely be eligible for GC after the scope exits if it was created in JavaScript-land, I think it may still be GC'd in this case but I'm not sure - maybe add some logging to the destructor to confirm it's cleaned up as expected?